### PR TITLE
Refactor into jcli-lib crate

### PR DESCRIPTION
**Pull request by filip-dulic-bloxico**
This is the beginning step for refactoring the jcli-lib crate out of the
jcli binary crate.

### CODE REVIEW

File: jcli-lib/src/address.rs [hunk 1]
```diff
@@ -1,11 +1,15 @@
...
+use crate::utils::key_parser::parse_pub_key;
 use chain_addr::{AddressReadable, Discrimination, Kind};
 use chain_crypto::{bech32::Bech32 as _, AsymmetricPublicKey, Ed25519, PublicKey};
+#[cfg(feature = "structopt")]
 use structopt::StructOpt;
 use thiserror::Error;
 
-#[derive(StructOpt)]
-#[structopt(name = "address", rename_all = "kebab-case")]
+#[cfg_attr(
```
Comment (danielSanchezQ) [inline jcli-lib/src/address.rs:8]: Ah!, nice to have 👍

**Review by eugene-babichenko — COMMENTED** [review]
The idea itself is good, BUT. IMO what is being done currently defies the whole purpose of moving functionality into the library. This library includes things that should be there (argument parsing) since they are going to be used only in `jcli` and this is not what I would call a good API either.

See my review comment below: instead of moving **everything** into the library we should start to gradually separate the functionality, that is needed to be shared to other apps.

I don't even mind not having the full functionality of `jcli` in the library in the first version, I understand this is a gigantic piece of work. This is perfectly fine as long as it is usable in other apps.

That said, instead of doing what we have done currently I would suggest to:

- Identify the current clients of this library (excluding **jcli**).
- Design a proper API that would not bring in the whole dependency tree of `jcli`. Given the nature of this library I assume that this would be just a bunch functions.

File: jcli-lib/src/address.rs [hunk 2]
```diff
@@ -1,11 +1,16 @@
...
```
[the remark this answers is not in the document]
Comment (danielSanchezQ) [reply jcli-lib/src/address.rs:9]: Actually, argument parsing could be reused too. Even inside jcli itself we have some duplicated code. Also some tools could reuse some of that code (it is being used for a tool now on catalyst-toolbox).

**Comment by danielSanchezQ** [comment]
> I don't even mind not having the full functionality of `jcli` in the library in the first version, I understand this is a gigantic piece of work. This is perfectly fine as long as it is usable in other apps.

+1


> That said, instead of doing what we have done currently I would suggest to:
> 
>     * Identify the current clients of this library (excluding **jcli**).
> 
>     * Design a proper API that would not bring in the whole dependency tree of `jcli`. Given the nature of this library I assume that this would be just a bunch functions.

IMO we could make every (needed) module could have an IO part (args parsing that can be reusable, some of the modules already had it, but they were private), and extract and refactor the `exec` methods into functions

File: jcli-lib/src/address.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (eugene-babichenko) [reply jcli-lib/src/address.rs:9]: > Even inside jcli itself we have some duplicated code

This is the problem of a client, not the problem of a library. ;)

> Also some tools could reuse some of that code

Need to look at the amount.

[the remark this answers is not in the document]
Comment (danielSanchezQ) [reply jcli-lib/src/address.rs:9]: > This is the problem of a client, not the problem of a library

In this case, it is both.

**Comment by eugene-babichenko** [comment]
> IMO we could make every (needed) module could have an IO part (args parsing that can be reusable, some of the modules already had it, but they were private), and extract and refactor the exec methods into functions

Sounds about right. I am still a bit skeptical about sharing `structopt` definitions, but if we really need them this is a cleaner approach. At least we can put all of them in a feature-gated module instead of playing around with `cfg_attr` which certainly doesn't contribute to readability.

**Comment by danielSanchezQ** [comment]
I let it as a draft for now, while we discuss what we will finally do.

File: jcli-lib/src/address.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (mzabaluev) [reply jcli-lib/src/address.rs:9]: The library could implement things like parsing common argument formats. Probably not the options and subcommand structures though, these should be defined in the CLI tool.

**Comment by mzabaluev** [comment]
jcli is currently used as a library in catalyst-toolbox, so it would make sense to check what is it used for and provide an API.

**Comment by minikin** [comment]
@filip-dulic-bloxico  Closing it. Feel free to reopen if we really need it.

**Pull request closed without merging**